library/std/src/sys/uefi/common.rs

```diff
@@ -0,0 +1,493 @@
+    let r = unsafe {
+        ((*boot_services.as_ptr()).install_protocol_interface)(
+            handle,
+            &mut guid,
+            r_efi::efi::NATIVE_INTERFACE,
+            interface as *mut T as *mut crate::ffi::c_void,
+        )
+    };
+    if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+}
+
+pub(crate) fn uninstall_protocol<T>(
+    handle: r_efi::efi::Handle,
+    mut guid: r_efi::efi::Guid,
+    interface: &mut T,
+) -> io::Result<()> {
+    let boot_services = get_boot_services().ok_or(BOOT_SERVICES_ERROR)?;
+    let r = unsafe {
+        ((*boot_services.as_ptr()).uninstall_protocol_interface)(
+            handle,
+            &mut guid,
+            interface as *mut T as *mut crate::ffi::c_void,
+        )
+    };
+    if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+}
+
+// A Helper trait for `ProtocolWrapper<P>`.
+pub(crate) trait Protocol {
+    const PROTOCOL_GUID: Guid;
+}
+
+// A wrapper for creating protocols from Rust in Local scope. Uninstalls the protocol on Drop.
+pub(crate) struct ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    handle: NonNull<crate::ffi::c_void>,
+    protocol: Box<P>,
+}
+
+impl<P> ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    #[inline]
+    pub(crate) fn new(handle: NonNull<crate::ffi::c_void>, protocol: Box<P>) -> Self {
+        Self { handle, protocol }
+    }
+
+    pub(crate) fn install_protocol(protocol: P) -> io::Result<ProtocolWrapper<P>> {
+        let mut handle: r_efi::efi::Handle = crate::ptr::null_mut();
+        let mut protocol = Box::new(protocol);
+        install_protocol::<P>(&mut handle, P::PROTOCOL_GUID, &mut protocol)?;
+        let handle = NonNull::new(handle)
+            .ok_or(io::const_io_error!(io::ErrorKind::Uncategorized, "found null handle"))?;
+        Ok(Self::new(handle, protocol))
+    }
+
+    pub(crate) fn install_protocol_in(
+        protocol: P,
+        mut handle: r_efi::efi::Handle,
+    ) -> io::Result<ProtocolWrapper<P>> {
+        let mut protocol = Box::new(protocol);
+        install_protocol::<P>(&mut handle, P::PROTOCOL_GUID, &mut protocol)?;
+        let handle = NonNull::new(handle)
+            .ok_or(io::const_io_error!(io::ErrorKind::Uncategorized, "found null handle"))?;
+        Ok(Self::new(handle, protocol))
+    }
+
+    #[inline]
+    pub(crate) fn handle(&self) -> NonNull<crate::ffi::c_void> {
+        self.handle
+    }
+}
+
+impl<P> Drop for ProtocolWrapper<P>
+where
+    P: Protocol,
+{
+    #[inline]
+    fn drop(&mut self) {
+        let _ = uninstall_protocol::<P>(self.handle.as_ptr(), P::PROTOCOL_GUID, &mut self.protocol);
+    }
+}
+
+/// Get the BootServices Pointer.
+pub(crate) fn get_boot_services() -> Option<NonNull<r_efi::efi::BootServices>> {
+    let system_table: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table().cast();
+    let boot_services = unsafe { (*system_table.as_ptr()).boot_services };
+    NonNull::new(boot_services)
+}
```

Comment:
I do not think that is the correct way of handling System Table and Image Handle. In fact, I am thinking of reverting the earlier changes regarding panicking in `system_table` and `image_handle` public functions. 
The problem is these functions are used in places where panic should not happen. Example in the allocation implementation, exit, abort, etc. 
I think it might be better to call unwrap at the site instead in places where it is fine to panic.